api/indexer/indexer.go

```diff
@@ -8,230 +8,230 @@ import (
+	case blockWindow == 0:
+		return nil, errZeroBlockWindow
 	}
 	blockDB, err := pebble.New(filepath.Join(path, "block"), pebble.NewDefaultConfig(), prometheus.NewRegistry())
 	if err != nil {
 		return nil, err
 	}
 
-	blockIDCache, err := cache.NewFIFO[ids.ID, uint64](int(blockWindow))
-	if err != nil {
-		return nil, err
-	}
-	blockCache, err := cache.NewFIFO[uint64, *chain.ExecutedBlock](int(blockWindow))
-	if err != nil {
-		return nil, err
-	}
 	i := &Indexer{
 		blockDB:            blockDB,
-		blockIDToHeight:    blockIDCache,
-		blockHeightToBlock: blockCache,
+		blockIDToHeight:    make(map[ids.ID]uint64, int(blockWindow)),
+		blockHeightToBlock: make(map[uint64]*chain.ExecutedBlock, int(blockWindow)),
+		txCache:            make(map[ids.ID]cachedTransaction),
 		blockWindow:        blockWindow,
 		parser:             parser,
-		txDB:               txDB,
+		lastHeight:         math.MaxUint64,
 	}
+
 	return i, i.initBlocks()
 }
 
 func (i *Indexer) initBlocks() error {
 	// Load blockID <-> height mapping
-	iter := i.blockDB.NewIterator()
+	iter := i.blockDB.NewIteratorWithPrefix(blockEntryKeyPrefix)
 	defer iter.Release()
 
-	var lastHeight uint64
 	for iter.Next() {
 		value := iter.Value()
 		blk, err := chain.UnmarshalExecutedBlock(value, i.parser)
 		if err != nil {
 			return err
 		}
-
-		i.blockIDToHeight.Put(blk.Block.GetID(), blk.Block.Hght)
-		i.blockHeightToBlock.Put(blk.Block.Hght, blk)
-		lastHeight = blk.Block.Hght
+		i.insertBlockIntoCache(blk)
 	}
 	if err := iter.Error(); err != nil {
 		return err
 	}
 	iter.Release()
 
-	i.lastHeight.Store(lastHeight)
-
-	if lastHeight > i.blockWindow {
-		lastRetainedHeight := lastHeight - i.blockWindow
-		lastRetainedHeightBytes := binary.BigEndian.AppendUint64(nil, lastRetainedHeight)
-		if err := i.blockDB.DeleteRange(nil, lastRetainedHeightBytes); err != nil {
+	if i.lastHeight > i.blockWindow {
+		lastRetainedHeight := i.lastHeight - i.blockWindow
+		lastRetainedHeightKey := blockEntryKey(lastRetainedHeight)
+		firstBlkKey := blockEntryKey(0)
+		if err := i.blockDB.DeleteRange(firstBlkKey, lastRetainedHeightKey); err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
 
 func (i *Indexer) Notify(_ context.Context, blk *chain.ExecutedBlock) error {
-	if err := i.storeTransactions(blk); err != nil {
-		return err
-	}
+	i.mu.Lock()
+	i.insertBlockIntoCache(blk)
+	i.mu.Unlock()
+
 	return i.storeBlock(blk)
 }
 
-func (i *Indexer) storeBlock(blk *chain.ExecutedBlock) error {
-	if i.blockWindow == 0 {
-		return nil
+// insertBlockIntoCache add the given block and its transactions to the
+// cache. It returns a slice of the evicted block heights.
+// assumes the write lock is held
+func (i *Indexer) insertBlockIntoCache(blk *chain.ExecutedBlock) {
+	if _, ok := i.blockHeightToBlock[blk.Block.Hght-i.blockWindow]; ok {
+		evictedBlockHeight := blk.Block.Hght - i.blockWindow
+
+		// find the block in the blocks cache.
+		evictedBlk := i.blockHeightToBlock[evictedBlockHeight]
```

Comment:
Can we avoid doing the same map lookup twice? We need this block, so we can just set this local variable in the if condition